arangod/Replication2/ReplicatedLog/LogFollower.cpp

```diff
@@ -204,65 +221,72 @@ auto replicated_log::LogFollower::appendEntries(AppendEntriesRequest req)
+       toBeResolved =
+           std::move(toBeResolved)](futures::Try<Result>&& tryRes) mutable
       -> std::pair<AppendEntriesResult, DeferredAction> {
     // We have to release the guard after this lambda is finished.
     // Otherwise it would be released when the lambda is destroyed, which
     // happens *after* the following thenValue calls have been executed. In
     // particular the lock is held until the end of the future chain is reached.
     // This will cause deadlocks.
-    decltype(selfGuard) self = std::move(selfGuard);
+    decltype(inFlightGuard) inFlightGuardLocal = std::move(inFlightGuard);
+    auto data = self->_guardedFollowerData.getLockedGuard(ADB_HERE);
+    // _appendEntriesInFlight = false;
 
     auto const& res = tryRes.get();
     {
       // This code block does not throw any exceptions. This is executed after
       // we wrote to the on-disk-log.
       static_assert(noexcept(res.fail()));
       if (res.fail()) {
-        LOG_CTX("216d8", ERR, self->_follower._loggerContext)
+        LOG_CTX("216d8", ERR, data->_follower._loggerContext)
             << "failed to insert log entries: " << res.errorMessage();
         return std::make_pair(
             AppendEntriesResult::withPersistenceError(
-                self->_follower._currentTerm, req.messageId, res),
+                data->_follower._currentTerm, req.messageId, res),
             DeferredAction{});
       }
 
       // commit the write in memory
       static_assert(
           std::is_nothrow_move_assignable_v<decltype(newInMemoryLog)>);
-      self->_inMemoryLog = std::move(newInMemoryLog);
+      data->_inMemoryLog = std::move(newInMemoryLog);
 
-      LOG_CTX("dd72d", TRACE, self->_follower._loggerContext)
+      LOG_CTX("dd72d", TRACE, data->_follower._loggerContext)
           << "appended " << req.entries.size() << " log entries after "
           << req.prevLogEntry.index
           << ", leader commit index = " << req.leaderCommit;
     }
 
-    auto action = self->checkCommitIndex(
+    auto action = data->checkCommitIndex(
         req.leaderCommit, req.lowestIndexToKeep, std::move(toBeResolved));
 
     static_assert(noexcept(AppendEntriesResult::withOk(
-        self->_follower._currentTerm, req.messageId)));
+        data->_follower._currentTerm, req.messageId)));
     static_assert(std::is_nothrow_move_constructible_v<DeferredAction>);
     return std::make_pair(AppendEntriesResult::withOk(
-                              self->_follower._currentTerm, req.messageId),
+                              data->_follower._currentTerm, req.messageId),
                           std::move(action));
   };
   static_assert(std::is_nothrow_move_constructible_v<
                 decltype(checkResultAndCommitIndex)>);
 
   // Action
-  return core->insertAsync(std::move(iter), req.waitForSync)
+  auto f = core->insertAsync(std::move(iter), req.waitForSync);
+  // Release mutex here, otherwise we might deadlock in
+  // checkResultAndCommitIndex
```

Comment:
```suggestion
  // Release mutex here, otherwise we might deadlock in
  // checkResultAndCommitIndex if another request arrives before the previous one was processed.
```